### PR TITLE
REF: make FrameApply less stateful, no self.results

### PR DESCRIPTION
**Pull request by jbrockmendel**
Passing `results` through the call chain instead of accessing it in each place as `self.results`  makes this code much easier to reason about.

### CODE REVIEW

File: pandas/core/apply.py [hunk 1]
```diff
@@ -18,6 +18,8 @@
 if TYPE_CHECKING:
     from pandas import DataFrame, Series, Index
 
+ResType = Dict[int, Any]
```
Comment (jreback) [inline pandas/core/apply.py]: ResultType maybe

File: pandas/core/apply.py [hunk 2]
```diff
@@ -107,8 +114,17 @@ def f(x):
 
         # results
         self.result = None
-        self.res_index = None
-        self.res_columns = None
+        self._res_index: Optional["Index"] = None
+
+    @property
+    def res_index(self) -> "Index":
+        if self._res_index is None:
```
Comment (jreback) [inline pandas/core/apply.py]: any reason to not make an assert?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/apply.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/apply.py]: That was what I tried first, but lines wrap prettier with the shorter name

File: pandas/core/apply.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/apply.py]: thatd be fine

**Review by jreback — CHANGES_REQUESTED** [review]
just that error check change, otherwise lgtm.

**Comment by jbrockmendel** [comment]
updated+green

**Review by jreback — APPROVED** [review]

**Pull request merged**